nautobot/ipam/views.py

```diff
@@ -804,6 +819,159 @@ def post(self, request):
+                    )
+                    vm_ip6 = list(
+                        VirtualMachine.objects.filter(primary_ip6__in=collapsed_ips).values_list("pk", flat=True)
+                    )
+
+                    ip_to_interface_assignments = []
+                    # Update IPAddress to Interface Assignments
+                    for assignment in IPAddressToInterface.objects.filter(ip_address__in=collapsed_ips):
+                        updated_attributes = model_to_dict(assignment)
+                        updated_attributes["ip_address"] = merged_ip
+                        updated_attributes["interface"] = Interface.objects.filter(
+                            pk=updated_attributes["interface"]
+                        ).first()
+                        updated_attributes["vm_interface"] = VMInterface.objects.filter(
+                            pk=updated_attributes["vm_interface"]
+                        ).first()
+                        ip_to_interface_assignments.append(updated_attributes)
+                    # Update Service m2m field with IPAddresses
+                    services = list(Service.objects.filter(ip_addresses__in=collapsed_ips).values_list("pk", flat=True))
+                    # Delete Collapsed IPs
+                    try:
+                        _, deleted_info = collapsed_ips.delete()
+                        deleted_count = deleted_info[IPAddress._meta.label]
+                    except ProtectedError as e:
+                        logger.info("Caught ProtectedError while attempting to delete objects")
+                        handle_protectederror(collapsed_ips, request, e)
+                        return redirect(self.get_return_url(request))
+                    msg = (
+                        f"Merged {deleted_count} {self.queryset.model._meta.verbose_name} "
+                        f'into <a href="{merged_ip.get_absolute_url()}">{escape(merged_ip)}</a>'
+                    )
+                    merged_ip.validated_save()
+                    # After some testing
+                    # We have to update the ForeignKey fields after merged_ip is saved to make the operation valid
+                    for assignment in ip_to_interface_assignments:
+                        IPAddressToInterface.objects.create(**assignment)
+                    # Update Device primary_ip fields of the Collapsed IPs
+                    Device.objects.filter(pk__in=device_ip4).update(primary_ip4=merged_ip)
+                    Device.objects.filter(pk__in=device_ip6).update(primary_ip6=merged_ip)
+                    VirtualMachine.objects.filter(pk__in=vm_ip4).update(primary_ip4=merged_ip)
+                    VirtualMachine.objects.filter(pk__in=vm_ip6).update(primary_ip6=merged_ip)
+                    for service in services:
+                        Service.objects.get(pk=service).ip_addresses.add(merged_ip)
+                    logger.info(msg)
```

Comment:
Not ideal to be logging HTML - consider using a different `msg` for the `logger.info` versus the `messages.success`